File: jvm/src/main/java/com/concord/wordcounter/WordSource.java
package com.concord;

import com.concord.*;
import com.concord.swift.*;
import java.util.*;

public class WordSource extends Computation {
  private final String[] words = {"foo", "bar", "baz", "fiz", "buzz"};
  private final Random rand = new Random();
  private String sample() {
    int idx = rand.nextInt(words.length);
    return words[idx];
  }

  public void init(ComputationContext ctx) {
    System.out.println("WordSource.java initialized");
    ctx.setTimer("loop", System.currentTimeMillis());
  }

  public void processRecord(ComputationContext ctx, Record record) {
    throw new RuntimeException("Method not implemented");
  }

  public void processTimer(ComputationContext ctx, String key, long time) {
    // Stream, key, value. Empty value, no need for val
    for (int i = 0; i < 1024; ++i) {
      ctx.produceRecord("words".getBytes(), this.sample().getBytes(),
                        "-".getBytes());
    }

    ctx.setTimer(key, System.currentTimeMillis() + 5000);
  }

  public Metadata metadata() {
    return new Metadata("word-source", new HashSet<StreamTuple>(),
                        new HashSet<String>(Arrays.asList("words")));
  }

  public static void main(String[] args) {
    ServeComputation.serve(new WordSource());
  }
}

File: scala/src/main/scala/com/concord/wordcounter/WordSource.scala
package com.concord.wordcounter

import java.util.{ HashSet => MutableHashSet}

import com.concord._
import com.concord.swift._

import scala.util.Random


class WordSource extends Computation {
  private val words = Array("foo", "bar", "baz", "fiz", "buzz")
  private val rand = new Random()

  def sample(): String = words( rand.nextInt(words.length) )

  override def init(ctx: ComputationContext): Unit = {
    println(s"${this.getClass.getSimpleName} initialized")
    ctx.setTimer("loop", System.currentTimeMillis())
  }

  override def processRecord(ctx: ComputationContext, record: Record): Unit = ???

  override def processTimer(ctx: ComputationContext, key: String, time: Long): Unit = {
    // Stream, key, value. Empty value, no need for val
    Range(0, 1024).foreach {
      i => ctx.produceRecord("words".getBytes, sample().getBytes, "-".getBytes)
    }

    ctx.setTimer(key, System.currentTimeMillis() + 5000)
  }

  override def metadata(): Metadata = {
    val ostreams = new MutableHashSet[String](java.util.Arrays.asList("words"))
    new Metadata("word-source", new MutableHashSet[StreamTuple](), ostreams)
  }

}


object WordSource {
  def main(args: Array[String]): Unit = {
    ServeComputation.serve(new WordSource())
  }
}
